### PR TITLE
Homework 3; PDP add cart

### PR DESCRIPTION
**Pull request by zborovskaya**

### CODE REVIEW

File: src/main/java/com/es/phoneshop/model/product/service/CartServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+import com.es.phoneshop.model.product.dao.ArrayListProductDao;
+
+import javax.servlet.http.HttpSession;
+import java.text.NumberFormat;
+import java.text.ParseException;
+import java.util.List;
+
+public class CartServiceImpl implements CartService {
+    private static volatile CartService instance;
+    private final static String NOT_NUMBER = "Not a number";
```
Comment (yan-piatkevich) [inline src/main/java/com/es/phoneshop/model/product/service/CartServiceImpl.java]: Constants generally declared at the top of the class state, check and fix for the entire project

File: src/main/java/com/es/phoneshop/model/product/service/CartServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    private final static String OUT_OF_STOCK = "Out of stock, available ";
```
Comment (yan-piatkevich) [inline src/main/java/com/es/phoneshop/model/product/service/CartServiceImpl.java]: Please align your modifiers order with jls - https://docs.oracle.com/javase/specs/jls/se7/html/jls-8.html#jls-8.3.1 
Check and fix for the entire project

File: src/main/java/com/es/phoneshop/model/product/service/ServiceException.java [hunk 3]
```diff
@@ -0,0 +1,19 @@
+package com.es.phoneshop.model.product.service;
+
+public class ServiceException extends Exception {
```
Comment (yan-piatkevich) [inline src/main/java/com/es/phoneshop/model/product/service/ServiceException.java]: ServiceException is too abstract name, and you should give appropriate names to sustain self-documentability of your code. Rename this one or add some subclasses with exact names.

File: src/main/java/com/es/phoneshop/web/ProductDetailsPageServlet.java [hunk 4]
```diff
@@ -1,29 +1,78 @@
 package com.es.phoneshop.web;
 
+import com.es.phoneshop.model.product.bean.Cart;
+import com.es.phoneshop.model.product.bean.RecentViewCart;
 import com.es.phoneshop.model.product.dao.ArrayListProductDao;
 import com.es.phoneshop.model.product.dao.ProductDao;
+import com.es.phoneshop.model.product.service.*;
```
Comment (yan-piatkevich) [inline src/main/java/com/es/phoneshop/web/ProductDetailsPageServlet.java:7]: Expand

File: src/main/java/com/es/phoneshop/web/ProductListPageServlet.java [hunk 5]
```diff
@@ -15,23 +15,29 @@
 
 public class ProductListPageServlet extends HttpServlet {
     private ProductDao productDao;
+    private final static String QUERY = "query";
+    private static final String PRODUCTS = "products";
+    private static final String SORT = "sort";
+    private static final String ORDER = "order";
+
```
Comment (yan-piatkevich) [inline src/main/java/com/es/phoneshop/web/ProductListPageServlet.java:22]: Redundant line

File: src/main/java/com/es/phoneshop/model/product/service/RecentViewedServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,51 @@
+package com.es.phoneshop.model.product.service;
+
+import com.es.phoneshop.model.product.bean.Product;
+import com.es.phoneshop.model.product.bean.RecentViewCart;
+import com.es.phoneshop.model.product.dao.ArrayListProductDao;
+
+import javax.servlet.http.HttpSession;
+import java.util.List;
+
+public class RecentViewedServiceImpl implements RecentViewedService {
```
Comment (yan-piatkevich) [inline src/main/java/com/es/phoneshop/model/product/service/RecentViewedServiceImpl.java:10]: Tests?

Comment (zborovskaya) [reply src/main/java/com/es/phoneshop/model/product/service/RecentViewedServiceImpl.java:10]: Test for this class?

**Review by yan-piatkevich — APPROVED** [review]

**Pull request merged**